scripts/bin/auto-install

```diff
@@ -73,92 +57,142 @@ rest_call() {
-            exit 1
-        fi
-    done
+    if ! echo "$_response" | grep -Iq '"status":"OK"'; then
+        log "ERROR: ACKNOWLEDGE RESULT: $_assert not acknowledged"
+        return 1
+    fi
 
-    install_snap "$_snaps"
+    log "ACKNOWLEDGE RESULT: $_assert acknowledged"
 }
 
 # Install the asserted snap using the snapd REST API
 install_snap() {
-    for _snap in "$@"; do
-        _snap_result="$SNAP_COMMON/${_snap##*/}_result.$_date"
-
-        # install the snap
-        rest_call install "$_snap_result" "$_snap"
-
-        if grep -Iq '"status":"Accepted"' "$_snap_result"; then
-            log "INSTALL RESULT: ${_snap##*/} installed"
-        else
-            log "ERROR: INSTALL RESULT: ${_snap##*/} not installed"
-            cp -f "${SNAP_COMMON}/auto-install.log" "$_mount_point"
-            exit 1
-        fi
+    _snap="$1"
 
-        track_channel "${_snap##*/}" "$_snap_result"
-    done
+    # install the snap
+    _response="$(rest_call install "$_snap")"
+
+    if ! echo "$_response" | grep -Iq '"status":"Accepted"'; then
+        log "ERROR: INSTALL RESULT: ${_snap##*/} not installed"
+        return 1
+    fi
+
+    log "INSTALL RESULT: ${_snap##*/} installed"
 }
 
 # set snapd to explicitly follow stable for a snap
-track_channel() {
-    # keep trying to set the channel to stable because it may take some time if a snap change is in process
-    # give up after 14 loops, where each loop sleeps twice as long as the previous
-    # where 14 loops takes 19.1 minutes
-    _snap_name=${1%%_*}
-    _log_name=$2
+track_stable() {
+    # Keep trying to set the channel to stable because it may take some
+    #   time if a snap change is in progress. Quadratically back off each loop,
+    #   and quit after 9 loops, for a sum total of ~17 minutes.
+
+    _snap_name="$1"
+
     _loop_count=0
     _sleep=2
 
-    while [ $_loop_count -lt 14 ]; do
-        sleep $_sleep
-        rest_call track "$_log_name" "$_snap_name"
+    while
+        _response="$(rest_call track "$_snap_name")"
 
-        if grep -Iq '"status":"Accepted"' "${_log_name}.channel"; then
+        if echo "$_response" | grep -Iq '"status":"Accepted"'; then
             log "${_snap_name} is now following stable"
-            cp -f "${SNAP_COMMON}/auto-install.log" "$_mount_point"
-            break
+            return 0
         fi
 
-        _loop_count=$((_loop_count+1))
-        _sleep=$((_sleep*2))
+        [ $_loop_count -lt 9 ]
+    do
+        sleep $_sleep
+
+        : $(( _loop_count += 1 ))
+        : $(( _sleep      *= 2 ))
     done
 
     log "WARN: failed to make $_snap_name follow stable"
     log "WARN: Manual intervention may be required for $_snap_name to refresh"
-    cp -f "${SNAP_COMMON}/auto-install.log" "$_mount_point"
+    return 1
+}
+
+eject_device() {
+    # Try to copy log file over
+    cp -f "$_log_file" "$_mount_point" || true
+
+    # Remove the temp directory
+    rm -r "$(dirname "$_log_file")"
```

Comment:
If we take my above comment about `$_log_file` we can change this to `rm -rf $_log_dir` - this is a reminder note :)